src/Aspire.Hosting/ContainerResourceBuilderExtensions.cs

```diff
@@ -699,6 +699,76 @@ public static IResourceBuilder<T> WithBuildSecret<T>(this IResourceBuilder<T> bu
+    /// <param name="builder">The resource builder for the container resource.</param>
+    /// <param name="destinationPath">The destination (absolute) path in the container.</param>
+    /// <param name="entries">The file system entries to create.</param>
+    /// <param name="defaultOwner">The default owner UID for the created or updated file system. Defaults to 0 for root.</param>
+    /// <param name="defaultGroup">The default group ID for the created or updated file system. Defaults to 0 for root.</param>
+    /// <param name="defaultMode">The default <see cref="UnixFileMode"/> ownership permissions for the created or updated file system. <see cref="UnixFileMode.None"/> will be treated as 0600 by DCP.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    /// <remarks>
+    /// <para>
+    /// For containers with a <see cref="ContainerLifetime.Persistent"/> lifetime, changing the contents of create file entries will result in the container being recreated.
+    /// Make sure any data being written to containers is idempotent for a given app model configuration. Specifically, be careful not to include any data that will be
+    /// unique on a per-run basis.
+    /// </para>
+    /// </remarks>
+    /// <example>
+    /// Create a directory called <c>custom-entry</c> in the container's file system at the path <c>/usr/data</c> and create a file called <c>entrypoint.sh</c> inside it with the content <c>echo hello world</c>.
+    /// The default permissions for these files will be for the user or group to be able to read and write to the files, but not execute them. entrypoint.sh will be created with execution permissions for the owner.
+    /// <code language="csharp">
+    /// var builder = DistributedApplication.CreateBuilder(args);
+    ///
+    /// builder.AddContainer("mycontainer", "myimage")
+    ///     .WithCreateFile("/usr/data", new()
+    ///     {
+    ///         new ContainerDirectory
+    ///         {
+    ///             Name = "custom-entry",
+    ///             Entries = new()
+    ///             {
+    ///                 new ContainerFile
+    ///                 {
+    ///                     Name = "entrypoint.sh",
+    ///                     Contents = "echo hello world",
+    ///                     Mode = UnixFileMode.UserExecute | UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.GroupRead | UnixFileMode.GroupWrite,
+    ///                 },
+    ///             },
+    ///         },
+    ///     },
+    ///     defaultOwner: 1000,
+    ///     defaultMode: UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.GroupRead | UnixFileMode.GroupWrite);
+    /// </code>
+    /// </example>
+    public static IResourceBuilder<T> WithCreateFile<T>(this IResourceBuilder<T> builder, string destinationPath, List<ContainerFileSystemItem> entries, int defaultOwner = 0, int defaultGroup = 0, UnixFileMode defaultMode = UnixFileMode.None) where T : ContainerResource
```

Comment:
`List<ContainerFileSystemItem>` is too constraining, I believe we would prefer IEnumerable in this case.

Reply:
Main reason I made it a List was to enable being lazy and instantiating the argument with new(), but that can easily be changed.